### PR TITLE
 New bidder : Added Preciso bidder

### PR DESCRIPTION
**Pull request by PrecisoSRL**
Added new bidder

### CODE REVIEW

**Comment by muuki88** [comment]
Docs: https://github.com/prebid/prebid.github.io/pull/4639

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            return Collections.emptyList();
+        }
+
+        return bidResponse.getSeatbid().stream().filter(Objects::nonNull)
+                .map(SeatBid::getBid).filter(Objects::nonNull)
+                .flatMap(Collection::stream).filter(Objects::nonNull)
+                .map(bid -> BidderBid.of(bid, getBidType(bid), bidResponse.getCur())).toList();
+    }
+
+    private ExtImpPreciso parseImpExt(Imp imp) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:105]: Please, place methods in call order.

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+
+        return bidResponse.getSeatbid().stream().filter(Objects::nonNull)
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:99]: Please, place intermediate stream operations on separate lines for better readability.

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                };
+    private static final String BIDDER_CURRENCY = "USD";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+    private final CurrencyConversionService currencyConversionService;
+
+    public PrecisoBidder(String endpointUrl,
+                         CurrencyConversionService currencyConversionService,
+                         JacksonMapper mapper) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:51]: Nitpick. Add empty line.

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        } catch (PreBidException e) {
+            throw new PreBidException(
+                    String.format(
+                    "Unable to convert provided bid floor currency from %s to %s for imp `%s`",
+                    bidFloorCur, BIDDER_CURRENCY, impId));
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, Price bidFloorPrice) {
+
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:146]: Nitpick. Redundant empty line.

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        return imp.toBuilder().bidfloorcur(ObjectUtil.getIfNotNull(bidFloorPrice, Price::getCurrency))
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:147]: Same here. Place builder method calls on separate lines.

File: src/main/proto/openrtb-xapi.proto [hunk 6]
```diff
@@ -0,0 +1,565 @@
+//
```
Comment (And1sS) [inline src/main/proto/openrtb-xapi.proto:1]: This should not be a part of repository. As proto files are downloaded as part of build process. Please remove.

File: src/main/proto/openrtb.proto [hunk 7]
```diff
@@ -0,0 +1,2113 @@
+syntax = "proto2";
```
Comment (And1sS) [inline src/main/proto/openrtb.proto:1]: Same here.

File: src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    private CurrencyConversionService currencyConversionService;
+
+    @Before
+    public void setUp() {
+        precisoBidder = new PrecisoBidder(ENDPOINT_URL, currencyConversionService, jacksonMapper);
+    }
+
+    @Test
+    public void creationShouldFailOnInvalidEndpointUrl() {
+
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java:60]: Redundant empty line at the beginning of each test. Please, remove and check for similar occurences.

**Review by And1sS — CHANGES_REQUESTED** [review]

File: src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java (hunk 8 above)
Comment (PrecisoSRL) [reply src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java:60]: added the changes as per your suggestion

File: src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+    private ExtImpPreciso parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), PRECISO_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, Price bidFloorPrice) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/preciso/PrecisoBidder.java:118]: Please, place methods in call order. 
PS. For example, this method is called from `makeHttpRequests`, so it should be placed after it.

File: src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java (hunk 8 above)
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java:60]: Still redundant empty line, also please check for similar occurences.

**Review by And1sS — DISMISSED** [review]

File: src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java (hunk 8 above)
Comment (PrecisoSRL) [reply src/test/java/org/prebid/server/bidder/preciso/PrecisoBidderTest.java:60]: By removing the empty line, we get the check style error on the maven build using mvn clean package, as shown below
image

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**